src/charm.py

```diff
@@ -0,0 +1,257 @@
+                        'PGID': '1000',  # copied from example docker run
+                        'TZ': 'Etc/UTC',  # copied from example docker run
+                    },
+                }
+            },
+        }
+
+    def _push_library_to_storage(self) -> None:
+        """Push user provided or default calibre-library resource to storage."""
+        library_write_behaviour, _ = self._get_library_write_behaviour()
+        if library_write_behaviour is None:
+            return
+        container = self.framework.model.unit.containers[CONTAINER_NAME]
+        if contents := container.list_files('/books/'):
+            msg_prefix = f'_push_library_to_storage: {library_write_behaviour=}: '
+            match library_write_behaviour:
+                case 'skip':
+                    logger.debug(msg_prefix + 'returning')
+                    return
+                case 'clean':
+                    logger.debug(msg_prefix + 'cleaning ...')
+                    for fileinfo in contents:
+                        container.remove_path(fileinfo.path, recursive=True)
+        logger.debug(f'_push_library_to_storage: {container=}')
+        library_path = Path(self.framework.model.resources.fetch('calibre-library'))
+        library = library_path.read_bytes()
+        # if the default library 'resource' was the starter library.zip
+        # instead of the empty empty.zip, this logic could be removed
+        if len(library):  # user provided library
+            logger.debug(f'_push_library_to_storage: {library_path=}')
+            self._push_and_extract_library(container, library)
+            return
+        # else: use default library
+        library_path = Path('.') / 'library.zip'
+        library = library_path.read_bytes()
+        assert library
+        logger.debug(f'_push_library_to_storage: {library_path=}')
+        self._push_and_extract_library(container, library)
+
+    def _get_library_write_behaviour(
+        self,
+    ) -> tuple[LibraryWriteBehaviour, ops.ActiveStatus] | tuple[None, ops.BlockedStatus]:
+        library_write_behaviour = self.model.config[LIBRARY_WRITE_CONFIG]
+        if library_write_behaviour not in LIBRARY_WRITE_BEHAVIOURS:
+            msg = f"invalid {LIBRARY_WRITE_CONFIG}: '{library_write_behaviour}'"
+            return None, ops.BlockedStatus(msg)
+        return cast(LibraryWriteBehaviour, library_write_behaviour), ops.ActiveStatus()
+
+    def _push_and_extract_library(self, container: ops.Container, library: bytes) -> None:
+        # use dtrx to extract library due to its consistent behaviour over many archive types
+        # could just add dtrx to the docker image? but it's somehow nice to use the off the shelf image
+        try:
+            container.exec(['which', 'dtrx']).wait()
+        except ops.pebble.ExecError:
+            logger.debug('_push_and_extract_library: installing dependencies')
+            container.exec(['apt', 'update']).wait()
```

Comment:
Same comment as earlier, that it seems like dependency installs would be better in pebble-ready.

Reply:
Definitely agree, thank you for this advice.